crates/numcodecs-python/src/schema.rs

```diff
@@ -0,0 +1,486 @@
+    let schema = schema.as_object()?;
+
+    let mut docs = String::new();
+
+    docs.push_str("# ");
+    docs.push_str(codec_id);
+
+    if let Some(Value::String(title)) = schema.get("title") {
+        docs.push_str(" (");
+        docs.push_str(title);
+        docs.push(')');
+    }
+
+    docs.push_str("\n\n");
+
+    if let Some(Value::String(description)) = schema.get("description") {
+        docs.push_str(description);
+        docs.push_str("\n\n");
+    }
+
+    docs.push_str("## Parameters\n\n");
+
+    for parameter in &parameters.named {
+        docs.push_str(" - ");
+        docs.push_str(parameter.name);
+
+        docs.push_str(" (");
+
+        if parameter.required {
+            docs.push_str("required");
+        } else {
+            docs.push_str("optional");
+        }
+
+        if let Some(default) = parameter.default {
+            docs.push_str(", default = `");
+            docs.push_str(&format!("{default}"));
+            docs.push('`');
+        }
+
+        docs.push(')');
+
+        if let Some(info) = &parameter.docs {
+            docs.push_str(": ");
+            docs.push_str(&info.replace('\n', "\n   "));
+        }
+
+        docs.push('\n');
+    }
+
+    if parameters.named.is_empty() {
+        if parameters.additional {
+            docs.push_str("This codec takes *any* parameters.");
+        } else {
+            docs.push_str("This codec does *not* take any parameters.");
+        }
+    } else if parameters.additional {
+        docs.push_str("\nThis codec takes *any* additional parameters.");
+    }
+
+    docs.truncate(docs.trim_end().len());
+
+    Some(docs)
+}
+
+pub fn signature_from_schema(schema: &Schema) -> String {
+    let parameters = parameters_from_schema(schema);
+
+    let mut signature = String::new();
+    signature.push_str("self");
+
+    for parameter in parameters.named {
+        signature.push_str(", ");
+        signature.push_str(parameter.name);
+
+        if let Some(default) = parameter.default {
+            signature.push('=');
+            signature.push_str(&format!("{default}"));
+        } else if !parameter.required {
+            signature.push_str("=None");
+        }
+    }
+
+    if parameters.additional {
+        signature.push_str(", **kwargs");
+    }
+
+    signature
+}
+
+#[allow(clippy::too_many_lines)] // FIXME
+fn parameters_from_schema(schema: &Schema) -> Parameters {
+    if schema.as_bool() == Some(true) {
+        return Parameters {
+            named: Vec::new(),
+            additional: true,
+        };
+    }
+
+    let Some(schema) = schema.as_object() else {
+        return Parameters {
+            named: Vec::new(),
+            additional: false,
+        };
+    };
+
+    let mut parameters = Vec::new();
+
+    let required = match schema.get("required") {
+        Some(Value::Array(required)) => &**required,
+        _ => &[],
+    };
+
+    if let Some(Value::Object(properties)) = schema.get("properties") {
```

Comment:
can this be extracted into a helper? add a comment: extract top-level properties